boa_gc/src/lib.rs

```diff
@@ -1,6 +1,448 @@
 //! Garbage collector for the Boa JavaScript engine.
 
-pub use gc::{
-    custom_trace, finalizer_safe, force_collect, unsafe_empty_trace, Finalize, Gc, GcCell as Cell,
-    GcCellRef as Ref, GcCellRefMut as RefMut, Trace,
-};
+#![allow(
+    clippy::let_unit_value,
+    clippy::should_implement_trait,
+    clippy::match_like_matches_macro,
+    clippy::new_ret_no_self,
+    // Putting the below on the allow list for now, but these should eventually be addressed
+    clippy::missing_safety_doc,
+    clippy::explicit_auto_deref,
+    clippy::borrow_deref_ref,
+)]
+
+use boa_profiler::Profiler;
+use std::cell::{Cell as StdCell, RefCell as StdRefCell};
+use std::mem;
+use std::ptr::NonNull;
+
+mod gc_box;
+mod internals;
+mod pointers;
+pub mod trace;
+
+pub use boa_gc_macros::{Finalize, Trace};
+
+pub use crate::trace::{Finalize, Trace};
+pub(crate) use gc_box::GcBox;
+pub use internals::{Ephemeron, GcCell as Cell, GcCellRef as Ref, GcCellRefMut as RefMut};
+pub use pointers::{Gc, WeakGc, WeakPair};
+
+pub type GcPointer = NonNull<GcBox<dyn Trace>>;
+
+thread_local!(pub static EPHEMERON_QUEUE: StdCell<Option<Vec<GcPointer>>> = StdCell::new(None));
+thread_local!(pub static GC_DROPPING: StdCell<bool> = StdCell::new(false));
+thread_local!(static BOA_GC: StdRefCell<BoaGc> = StdRefCell::new( BoaGc {
+    config: GcConfig::default(),
+    runtime: GcRuntimeData::default(),
+    adult_start: StdCell::new(None),
+}));
+
+struct GcConfig {
+    adult_threshold: usize,
+    growth_ratio: f64,
+}
+
+// Setting the defaults to an arbitrary value currently.
+//
+// TODO: Add a configure later
+impl Default for GcConfig {
+    fn default() -> Self {
+        Self {
+            adult_threshold: 1024,
+            growth_ratio: 0.8,
+        }
+    }
+}
+
+#[derive(Default)]
+struct GcRuntimeData {
+    collections: usize,
+    total_bytes_allocated: usize,
+    adult_bytes: usize,
+}
+
+struct BoaGc {
+    config: GcConfig,
+    runtime: GcRuntimeData,
+    adult_start: StdCell<Option<GcPointer>>,
+}
+
+impl Drop for BoaGc {
+    fn drop(&mut self) {
+        unsafe {
+            Collector::dump(self);
+        }
+    }
+}
+// Whether or not the thread is currently in the sweep phase of garbage collection.
+// During this phase, attempts to dereference a `Gc<T>` pointer will trigger a panic.
+
+struct DropGuard;
+
+impl DropGuard {
+    fn new() -> DropGuard {
+        GC_DROPPING.with(|dropping| dropping.set(true));
+        DropGuard
+    }
+}
+
+impl Drop for DropGuard {
+    fn drop(&mut self) {
+        GC_DROPPING.with(|dropping| dropping.set(false));
+    }
+}
+
+pub fn finalizer_safe() -> bool {
+    GC_DROPPING.with(|dropping| !dropping.get())
+}
+
+/// The GcAllocater handles initialization and allocation of garbage collected values.
+///
+/// The allocator can trigger a garbage collection
+pub struct BoaAlloc;
```

Comment:
I'm not sure why do we need this structure to construct the pointers. I think following `gc`'s APIs looks better and is easier to use and compose. Maybe I'm overlooking something?

Reply:
My thought behind this primarily was to take the heap allocation and gc trigger out of `GcBox::new()` heap allocation from the structs.

I might have leaned in a bit too heavily to the idea. Maybe it would be better to call this struct from `Gc::new()` and have a `Gc::new_cell()`.